### PR TITLE
ceph: placement in case of both PVC and non-PVC's 

### LINKED ISSUE #7176: Placement of device set OSDs are merged with the .spec.resources.osd
Seems like https://github.com/rook/rook/pull/6918 introduced a regression when using both PVC OSDs and raw device OSDs in the same ceph cluster, which was added in https://github.com/rook/rook/pull/6822

If the raw device OSDs and the PVC OSDs have different resource requests and placement config, the .spec.resources.osd merges with what's defined at the deviceset placement spec.

cluster spec to illustrate the problem
```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  dataDirHostPath: /var/lib/rook
  skipUpgradeChecks: false 
  continueUpgradeAfterChecksEvenIfNotHealthy: true
  # removeOSDsIfOutAndSafeToRemove: true
  mon:
    count: 3
    allowMultiplePerNode: false
    volumeClaimTemplate:
      spec:
        storageClassName: ebs-gp2
        resources:
          requests:
            storage: 10Gi
  mgr:
    modules:
    - name: pg_autoscaler
      enabled: true 
    - name: balancer
      enabled: true                  
  placement:
    mon:
      podAntiAffinity:   
        preferredDuringSchedulingIgnoredDuringExecution:        
        - weight: 100
          podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - rook-ceph-mon
            topologyKey: failure-domain.beta.kubernetes.io/zone
    osd:
      tolerations:
      - key: k8s.gdo.aws/role
        operator: Equal
        value: storage-ephemeral
        effect: NoSchedule
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: kops.k8s.io/instancegroup
              operator: In
              values:
              - storage-a-ephemeral 
              - storage-b-ephemeral
              - storage-c-ephemeral
    prepareosd:
      tolerations:
      - key: k8s.gdo.aws/role
        operator: Equal
        value: storage-ephemeral
        effect: NoSchedule
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: kops.k8s.io/instancegroup
              operator: In
              values:
              - storage-a-ephemeral 
              - storage-b-ephemeral
              - storage-c-ephemeral
  resources:
    osd:
      limits:
        cpu: "3500m"
        memory: "25Gi"
      requests:
        cpu: "2500m"
        memory: "12Gi"
    prepareosd:
      limits:
        cpu: "500m"
        memory: "2Gi"
      requests:
        cpu: "200m"
        memory: "1Gi"
  cephVersion:
    image: ceph/ceph:v15.2.4
    allowUnsupported: false
  dashboard:
    enabled: true
    urlPrefix: /ceph-dashboard
    port: 8443
    ssl: false
  monitoring:
    # requires Prometheus to be pre-installed
    enabled: true
    # namespace to deploy prometheusRule in. If empty, namespace of the cluster will be used.
    # Recommended:
    # If you have a single rook-ceph cluster, set the rulesNamespace to the same namespace as the cluster or keep it empty.
    # If you have multiple rook-ceph clusters in the same k8s cluster, choose the same namespace (ideally, namespace with prometheus
    # deployed) to set rulesNamespace for all the clusters. Otherwise, you will get duplicate alerts with multiple alert definitions.
    rulesNamespace: rook-ceph    
  network:
    hostNetwork: false
  crashCollector:
    disable: false
  # healthChecks
  # Valid values for daemons are 'mon', 'osd', 'status'
  healthCheck:
    daemonHealth:
      mon:
        disabled: false
        interval: 45s
      osd:
        disabled: false
        interval: 60s
      status:
        disabled: false
        interval: 60s
    # Change pod liveness probe, it works for all mon,mgr,osd daemons
    livenessProbe:
      mon:
        disabled: false
      mgr:
        disabled: false
      osd:
        disabled: false  
  storage:
    topologyAware: true
    useAllNodes: true
    useAllDevices: true
    config:
      # osdsPerDevice: "1"
      deviceClass: nvme
    storageClassDeviceSets:
    - name: zone-a
      count: 1
      portable: true
      tuneDeviceClass: true
      resources:
        limits:
          cpu: "3500m"
          memory: "7Gi"
        requests:
          cpu: "3000m"
          memory: "6.5Gi"
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: k8s.gdo.aws/zone
                operator: In
                values:
                - a 
              - key: k8s.gdo.aws/role
                operator: In
                values:
                - storage   
        tolerations:
        - key: k8s.gdo.aws/role
          operator: Equal
          value: storage
          effect: NoSchedule                
      volumeClaimTemplates:
      - metadata:
          name: data
          annotations:
            crushDeviceClass: ssd
        spec:
          resources:
            requests:
              storage: 305Gi
          storageClassName: ebs-gp2
          volumeMode: Block
          accessModes:
            - ReadWriteOnce
    - name: zone-b
      count: 1
      portable: true
      tuneDeviceClass: true
      resources:
        limits:
          cpu: "3500m"
          memory: "7Gi"
        requests:
          cpu: "3000m"
          memory: "6.5Gi"
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: k8s.gdo.aws/zone
                operator: In
                values:
                - b
              - key: k8s.gdo.aws/role
                operator: In
                values:
                - storage  
        tolerations:
        - key: k8s.gdo.aws/role
          operator: Equal
          value: storage
          effect: NoSchedule                
      volumeClaimTemplates:
      - metadata:
          name: data
          annotations:
            crushDeviceClass: ssd
        spec:
          resources:
            requests:
              storage: 305Gi
          storageClassName: ebs-gp2
          volumeMode: Block
          accessModes:
            - ReadWriteOnce 
    - name: zone-c
      count: 1
      portable: true
      tuneDeviceClass: true
      resources:
        limits:
          cpu: "3500m"
          memory: "7Gi"
        requests:
          cpu: "3000m"
          memory: "6.5Gi"
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: k8s.gdo.aws/zone
                operator: In
                values:
                - c
              - key: k8s.gdo.aws/role
                operator: In
                values:
                - storage 
        tolerations:
        - key: k8s.gdo.aws/role
          operator: Equal
          value: storage
          effect: NoSchedule
      volumeClaimTemplates:
      - metadata:
          name: data
          annotations:
            crushDeviceClass: ssd
        spec:
          resources:
            requests:
              storage: 305Gi
          storageClassName: ebs-gp2
          volumeMode: Block
          accessModes:
            - ReadWriteOnce
```

The rendered pvc OSD deployment for deviceset `zone-a` ends up with this affinity rules, which will not satisfy any of the nodes in the cluster. 

deployment for DeviceSet `zone-a`
```
  spec:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: k8s.gdo.aws/zone
              operator: In
              values:
              - a
            - key: k8s.gdo.aws/role
              operator: In
              values:
              - storage
            - key: kops.k8s.io/instancegroup
              operator: In
              values:
              - storage-a-ephemeral
              - storage-b-ephemeral
              - storage-c-ephemeral
```

deployment for DeviceSet `zone-b`
```
  spec:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: k8s.gdo.aws/zone
              operator: In
              values:
              - b
            - key: k8s.gdo.aws/role
              operator: In
              values:
              - storage
            - key: kops.k8s.io/instancegroup
              operator: In
              values:
              - storage-a-ephemeral
              - storage-b-ephemeral
              - storage-c-ephemeral
```

deployment for DeviceSet `zone-c`
```
  spec:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: k8s.gdo.aws/zone
              operator: In
              values:
              - c
            - key: k8s.gdo.aws/role
              operator: In
              values:
              - storage
            - key: kops.k8s.io/instancegroup
              operator: In
              values:
              - storage-a-ephemeral
              - storage-b-ephemeral
              - storage-c-ephemeral
```

### PR DESCRIPTION
**Pull request by subhamkrai**
**Description of your changes:**

In the case of PVC,
We'll first apply `placement.All()` and then we'll
apply `devicSet Placement`. If both are specified 
`placement.All` and `devicSet Placement` then we merging
both at least the `nodeAffinity`.

In case of non-PVC,
we apply `spec.placement`.

Signed-off-by: subhamkrai <srai@redhat.com>
co-authored-by: Travis Nielsen <tnielsen@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #7176 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/osd/provision_spec.go [hunk 1]
```diff
@@ -156,9 +158,15 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
 	// we are passing false in case of osd and prepare pods because we don't want to overlap placement of osd on PVC's and non-PVC's.
 	p := cephv1.GetOSDPlacement(c.spec.Placement)
 	if osdProps.onPVC() {
+		// apply storageClassDeviceSet PreparePlacement
 		osdProps.getPreparePlacement().ApplyToPodSpec(&podSpec, false)
+		// apply placement.All() if it is not empty
+		if !reflect.DeepEqual(c.spec.Placement.All(), rookv1.Placement{}) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/provision_spec.go:164]: Isn't `ApplyToPodSpec()` already a no-op if the placement is empty? I think you could remove this check. Or did something not work without it?

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 2]
```diff
@@ -658,9 +659,15 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 	// we are passing false in case of osd and prepare pods because we don't want to overlap placement of osd on PVC's and non-PVC's.
 	p := cephv1.GetOSDPlacement(c.spec.Placement)
 	if osdProps.onPVC() {
+		// apply storageClassDeviceSet Placement
 		osdProps.placement.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)
+		// apply placement.All() if it is not empty
+		if !reflect.DeepEqual(c.spec.Placement.All(), rookv1.Placement{}) {
+			p.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go:666]: Wait, I'm not seeing how this works. If both `all` and `osd` sections are specified in the placement, they would both be applied here. But we only want to apply the `all`, and **not** the `osd` section.

Comment (travisn) [reply pkg/operator/ceph/cluster/osd/spec.go:666]: I still feel like this is the right approach. Did this not work? And you could delete line 660 for `GetOSDPlacement()`.
```go
if osdProps.onPVC() {
	// Merge the "all" and storageClassDeviceSet placement
	p := c.spec.Placement.All().Merge(*osdProps.preparePlacement)
	p.ApplyToPodSpec(&podSpec, false)
} else {
	// Merge the "all" and "osd" placement
	p := cephv1.GetOSDPlacement(c.spec.Placement)
	p.ApplyToPodSpec(&podSpec, false)
}
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by subhamkrai** [comment]
with the suggestion we have 2 problem: 
1: in `c.spec.Placement.All().Merge(*osdProps.preparePlacement)`  `Merge()` it will replace  `c.spec.Placement.All()` with `*osdProps.preparePlacement` which is just opposite right, we want to replace `*osdProps.preparePlacement`. We can do this by swapping like `&osdProps.preparePlacement.Merge(c.spec.Placement.All())`but it shows error 
2: for `*osdProps.preparePlacement` we have a problem with the pointer that if `storageClassDeviceSet preparePlacement:` is not specified it will fail with `nill pointer` for this I can add a new check but it is not working.

**Comment by subhamkrai** [comment]
with this 
```
if osdProps.onPVC() {
		// apply storageClassDeviceSet Placement
		osdProps.placement.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)
		// apply placement.All() if it is not empty
		if !reflect.DeepEqual(c.spec.Placement.All(), rookv1.Placement{}) {
			p.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)
		}
	} else {
		p.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)
	}
```
question was:
> I'm not seeing how this works. If both `all` and `osd` sections are specified in the placement, they would both be applied here. But we only want to apply the `all`, and **not** the `osd` section.

It will not apply for both  `all` and `osd` we are passing `false` in `ApplyToPodSpec()` which make sure we are not merging . First with this `osdProps.placement.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)` `osdProps.placement`)(storageDeviceClassSet placement) will apply then we are checking for
```
if !reflect.DeepEqual(c.spec.Placement.All(), rookv1.Placement{}) 
``` 
with this we are checking if Placement.All() is not empty the we will apply `p.ApplyToPodSpec(&deployment.Spec.Template.Spec, false)`(p := cephv1.GetOSDPlacement(c.spec.Placement)).

Let me know if I'm not clear. Thanks

**Comment by travisn** [comment]
Ok I did some experimenting with the approach and ended up with a combination of our solutions. :) See [this commit](https://github.com/rook/rook/commit/6b94ecc3593b5d30460a7ccf00f5c9526a5077bc) for the details.
- We call `ApplyToPodSpec()` twice, first for "all" and second for the device set placement. This will allow the nodeAffinity to be merged if specified in both of them, and also give higher priority to the device set placement in case there is something specified in both places.
- No need for the `merge` flag on `ApplyToPodSpec()`. Now we just always merge the node affinity since that anti-merge isn't needed for the OSDs anymore.

If this works for you, go ahead and squash in your PR, thanks.

File: pkg/apis/rook.io/v1/placement.go [hunk 3]
```diff
@@ -45,19 +45,17 @@ func (p Placement) ApplyToPodSpec(t *v1.PodSpec, merge bool) {
...
 		return result
 	}
 
 	// merge the preferred node affinity that was already specified, and the placement's nodeAffinity
-	if merge {
-		result.PreferredDuringSchedulingIgnoredDuringExecution = append(
-			nodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution,
-			p.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution...)
-	}
+	result.PreferredDuringSchedulingIgnoredDuringExecution = append(
```
Comment (sp98) [inline pkg/apis/rook.io/v1/placement.go:56]: with this new change, won't be  merge `PreferredDuringSchedulingIgnoredDuringExecution`  from both `All` and `StorageClassDeviceSets`.  And end up having similar scenario mentioned in the original issue description but for `PreferredDuringSchedulingIgnoredDuringExecution`

Comment (travisn) [reply pkg/apis/rook.io/v1/placement.go:56]: I believe the original concern was that "osd" placement should not ever be used for the storageClassDeviceSet placement, and vice versa. Merging the affinity with "all" for both types of OSDs seems like a good thing. Or which comment in the original issue are you referring to? I could certainly still be missing something.

**Review by travisn — APPROVED** [review]
My testing looks good with a cluster on PVCs and non-PVCs. The storageClassDeviceSet was never merged with the placement from the `placement.osd` section, only with the `placement.all`.

**Pull request merged**